src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/ReadAndUpdateOperation.java

```diff
@@ -138,42 +138,52 @@ private Uni<UpdatedDocument> processUpdate(
       ReadDocument document, QueryExecutor queryExecutor, AtomicInteger modifiedCount) {
     return Uni.createFrom()
         .item(document)
-        // Perform update operation and save only if data is modified.
-        .onItem()
-        .transformToUni(
+
+        // perform update operation and save only if data is modified.
+        .flatMap(
             readDocument -> {
+
+              // if there is no document return null item
               if (readDocument == null) {
                 return Uni.createFrom().nullItem();
               }
-              final JsonNode originalDocument =
-                  readDocument.txnId() == null ? null : readDocument.document();
-              final boolean isInsert = (originalDocument == null);
-              // Apply document updates
+
+              // upsert if we have no transaction if before
+              boolean upsert = readDocument.txnId() == null;
+              JsonNode originalDocument = upsert ? null : readDocument.document();
+
+              // apply document updates
+              // if no changes return null item
               DocumentUpdater.DocumentUpdaterResponse documentUpdaterResponse =
-                  documentUpdater().applyUpdates(readDocument.document().deepCopy(), isInsert);
-              JsonNode updatedDocument = documentUpdaterResponse.document();
-              Uni<DocumentId> updated = Uni.createFrom().nullItem();
-              final WritableShreddedDocument writableShreddedDocument;
-              if (documentUpdaterResponse.modified()) {
-                writableShreddedDocument = shredder().shred(updatedDocument, readDocument.txnId());
-                updated = updatedDocument(queryExecutor, writableShreddedDocument);
-              } else {
-                writableShreddedDocument = null;
+                  documentUpdater().applyUpdates(readDocument.document().deepCopy(), upsert);
+              if (!documentUpdaterResponse.modified()) {
+                return Uni.createFrom().nullItem();
```

Comment:
We need to do null check on responses collected as list and passed to UpdateOperationPage

Reply:
Apologies, didn't realize its going into collect() that should handle null value.

Reply:
I did not change at all the previous logic :)